File: chapter-11-greedy/만들-수-없는-금액-book.py
import time

'''
1 1 2 3 9 

target = 1
x = 1
1 < 1
target += x

target = 2
x = 1
2 < 1
target += x

target = 3
x = 2
3 < 2
taarget += x

target = 5
x = 3
5 < 3
target += x

target = 8
x = 9
8 < 9

target = 8
'''

def solution(N, data):
    data.sort()

    result = 1
    for x in data:
        # 만들 수 없는 금액을 찾았을 때 반복 종료
        if result < x:
            break
        result += x

    return result


N = int(input())
data = list(map(int, input().split()))
start_time = time.time()
result = solution(N,data)
end_time = time.time()

print('실행 결과:', result)
print('걸린 시간:', end_time - start_time)

File: chapter-11-greedy/만들-수-없는-금액-book.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int N, vector<int> arr){
    sort(arr.begin(), arr.end());

    int target = 1;
    for (int i = 0; i < N; ++i) {
        // 만들 수 없는 금액을 찾았을 때 반복 종료
        if (target < arr[i]) break;
        target += arr[i];
    }

    return target;
} 

int main(){
    int N;
    cin >> N;

    vector<int> arr;
    for (int i = 0; i < N; ++i) {
        int x;
        cin >> x;
        arr.push_back(x);
    }

    int result = solution(N, arr);
    cout << "실행 결과: " << result << endl;
    return 0;
}
